### DISCUSSION TITLE
New to floorplan, struggling with customizing

### DISCUSSION THREAD

**Discussion by Maxxell** [ExperienceLovelace/ha-floorplan#87] [Q&A] [answered]
Hi!
I'm running HA on a Pi4.  
I think I came across an older version of floorplan first, so I started down the wrong path.  But now I am able to get a view of the example home.svg.  Then I was able to swap in my custom floorplan svg.  So far, so good.
But when I then import another .svg and name that the same as my lightbulb entity, thinks start to misbehave.  I can get the bulb to work (show the right icons based on the state of the bulb; and click the bulb to send a toggle command); but the image of the bulb moves to the top-left corner. 

Right now, to make changes, I go to my custom dash, click on the floorplan name at the top, then click edit and then "Raw configuration editor".  (Question #1: is this floorplan.yaml I see referenced around this topic?)

Here's the last 30+lines showing the portion of the template code provided that I've kept:
  - title: Floorplan JSG
    path: floorplan-jsg
    theme: ''
    badges: []
    cards:
      - type: vertical-stack
        cards:
          - type: horizontal-stack
            cards:
              - type: 'custom:floorplan-card'
                full_height: true
                config:
                  image: /local/floorplan/examples/home/JSGhome10.svg
                  stylesheet: /local/floorplan/examples/home/home.css
                  log_level: info
                  console_log_level: info
                  defaults:
                    tap_action: more-info
                  rules:
                    - element: light.68074866dc4f22afcc90
                      entity: light.68074866dc4f22afcc90
                      state_action:
                        action: call-service
                        service: floorplan.image_set
                        service_data: >-
                          /local/floorplan/examples/home/light_${entity.state}.svg
                      tap_action: toggle

Any ideas what I am missing?

**Comment by OzGav** [comment, marked as answer]
I'm on my phone but if you have a look at the other discussions you will see others with your same issue. The best way to do this is rather than inserting SVG at runtime is to put the bulb SVG in your base image and then use CSS styles to show or hide it

**Reply by Maxxell** [reply]
I do have the bulb (in the off position) in the SVG (i put it there in Inkscape by 'importing' the one i found in the examples files into the custom .svg that is my actual layout).  

But CSS styles are something I dont really know much about.  I dont see the bulb referenced in the home.css in the example so I didn't realize that was how to make the bulb image stay put.  When you get a chance, could you highlight the css needed to accomplish this?  It sounds like you're saying I should import 2 images (on and off) for each entity/bulb into the background svg file and then use css to toggle the entity/bulb image appropriately.  Am I thinking about this the right way?

**Reply by Maxxell** [reply]
SOLVED!
It had nothing to do with css.

It was the way I was importing the image in Inkscape.  I found a comment elsewhere that mentioned: 
>Also you need to make sure that the placeholder is placed directly within the svg (e.g. not in a layer in inkscape) or else the calculated coordinates will be wrong.

So I went back to my custom svg and at the import step, chose a different option.  Maybe I missed it, but nothing I've read so far mentions the importance of importing the SVG as "EMBED" instead of the default "Include" or "Link" when you do it in Inkscape.

**Reply by OzGav** [reply]
If you look at your code above you are using the image_set service which is placing a new image on your floorplan. You need to use the style_set call and just show or hide the image. You should have an off image and then on top of that have the on image in your base image. You then show or hide the on image and that will give you the effect of the bulb icon changing from on to off. Have a look at the other discussions  where this has been detailed I am pretty sure. If you can’t find it let me know.

**Reply by Maxxell** [reply]
That's so weird that you double-down on the css solution.  When I did the 'embed' option at the import the icon step, the bulb is working exactly as hoped.

Is this a situation where there are 2 acceptable options or am I going to regret it if I continue using this system for the rest of the bulbs in my layout?

**Reply by OzGav** [reply]
Yes i think it can work but I don’t do it that way and others have had issues like you. Have a look at this post https://community.home-assistant.io/t/floorplan-now-available-as-a-lovelace-card/115489/416?u=ozgav

**Reply by OzGav** [reply]
Glad you have it working though. Many ways to skin the cat!

**Reply by Maxxell** [reply]
Can you clarify something for me?

>You should have an off image and then on top of that have the on image in your base image

How would this work?  Inkspace yells at me if I try to have two objects with the same name in the master file.  How do I have an image for off and an image for on if I can only have 1?

**Reply by exetico** [reply]
Floorplan supports replacements of graphical elements. Check for docs, or some of the other discussions. There's a few examples, if I remember correctly :-)

**Reply by Maxxell** [reply]
If I understand you correctly, then it's not something you do in the base image.  The base image shows your floorplan and placeholders for active elements, and then you let Floorplan drop in new graphical elements that are stored elsewhere on top when needed.  

In my head, the image that shows up in the base image is just a placeholder, and its still there (just hidden behind the 'active' image) when floorplan displays a new image there because of a state call.  But what happens when you have different sizes/shapes of image for different states of the same entity (like a garage door or something)?  

What happens when floorplan is asked to swap in an image that is different size of the image it's replacing?  Does that make sense?

**Reply by OzGav** [reply]
> How would this work? Inkspace yells at me if I try to have two objects with the same name in the master file. How do I have an image for off and an image for on if I can only have 1?

For example, place the off image in Inkscape but do NOT add an ID to this as you won’t be changing anything about this object. Then place the on image on top and add the HA entity ID to this object. 

Now in the rules you will control the on image. If you hide the on image then the off image will be visible. If you show the on image it will obscure the off image and only it will be visible.

**Reply by OzGav** [reply]
> What happens when floorplan is asked to swap in an image that is different size of the image it's replacing? Does that make sense?

It does make sense. I think it will get sized or distorted to fit. Just try it and see what happens. If I was trying to put a rectangular shape in a square placeholder  then I would actually make the rectangular image square by adding transparent space around it.
